File: languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/SessionJsonImplTest.java
package com.godaddy.asherah.appencryption;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import org.json.JSONObject;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryption;
import com.godaddy.asherah.appencryption.utils.Json;

@ExtendWith(MockitoExtension.class)
class SessionJsonImplTest {

  @Mock
  EnvelopeEncryption<String> envelopeEncryption;
  @InjectMocks
  SessionJsonImpl<String> sessionJsonImpl;

  @Test
  void testConstructor() {
    Session<?, ?> session = new SessionJsonImpl<>(envelopeEncryption);
    assertNotNull(session);
  }

  @Test
  void testDecrypt() {
    JSONObject expectedJson = new JSONObject("{\"some_key\": 123}");
    byte[] utf8Bytes = new Json(expectedJson).toUtf8();
    when(envelopeEncryption.decryptDataRowRecord(any())).thenReturn(utf8Bytes);

    JSONObject actualJson = sessionJsonImpl.decrypt("some data row record");
    assertTrue(expectedJson.similar(actualJson));
  }

  @Test
  void testEncrypt() {
    String expectedDataRowRecord = "some data row record";
    when(envelopeEncryption.encryptPayload(any())).thenReturn(expectedDataRowRecord);

    String actualDataRowRecord = sessionJsonImpl.encrypt(new JSONObject("{\"some_key\": 123}"));
    assertEquals(expectedDataRowRecord, actualDataRowRecord);
  }

  @Test
  void testCloseSuccess() {
    sessionJsonImpl.close();

    // Verify proper resources are closed
    verify(envelopeEncryption).close();
  }

  @Test
  void testCloseWithCloseFailShouldReturn() {
    doThrow(RuntimeException.class).when(envelopeEncryption).close();
    sessionJsonImpl.close();

    verify(envelopeEncryption).close();
  }

}

File: csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionJsonImplTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.Envelope;
using Microsoft.Extensions.Logging;
using Moq;
using Newtonsoft.Json.Linq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption
{
    public class SessionJsonImplTest : IDisposable
    {
        private readonly Mock<IEnvelopeEncryption<string>> envelopeEncryptionMock;
        private readonly Mock<ILogger> mockLogger;
        private readonly SessionJsonImpl<string> sessionJsonImpl;

        public SessionJsonImplTest()
        {
            envelopeEncryptionMock = new Mock<IEnvelopeEncryption<string>>();
            mockLogger = new Mock<ILogger>();
            sessionJsonImpl = new SessionJsonImpl<string>(envelopeEncryptionMock.Object, mockLogger.Object);
        }

        [Fact]
        public void TestConstructor()
        {
            SessionJsonImpl<string> session = new SessionJsonImpl<string>(envelopeEncryptionMock.Object, mockLogger.Object);
            Assert.NotNull(session);
        }

        [Fact]
        public void TestDecrypt()
        {
            const string json = @"{key:'some_key', value:123}";
            JObject expectedJson = JObject.Parse(json);
            byte[] utf8Bytes = new Asherah.AppEncryption.Util.Json(expectedJson).ToUtf8();

            envelopeEncryptionMock.Setup(x => x.DecryptDataRowRecord(It.IsAny<string>())).Returns(utf8Bytes);

            JObject actualJson = sessionJsonImpl.Decrypt("some data row record");
            Assert.Equal(expectedJson, actualJson);
        }

        [Fact]
        public void TestEncrypt()
        {
            const string expectedDataRowRecord = "some data row record";
            const string json = @"{key:'some_key', value:123}";
            JObject jObject = JObject.Parse(json);

            envelopeEncryptionMock.Setup(x => x.EncryptPayload(It.IsAny<byte[]>())).Returns(expectedDataRowRecord);

            string actualDataRowRecord = sessionJsonImpl.Encrypt(jObject);
            Assert.Equal(expectedDataRowRecord, actualDataRowRecord);
        }

        [Fact]
        public void TestDispose()
        {
            sessionJsonImpl.Dispose();

            // Verify proper resources are closed
            envelopeEncryptionMock.Verify(x => x.Dispose());
        }

        [Fact]
        public void TestCloseWithCloseFailShouldReturn()
        {
            envelopeEncryptionMock.Setup(x => x.Dispose()).Throws<Exception>();
            sessionJsonImpl.Dispose();
            envelopeEncryptionMock.Verify(x => x.Dispose());
        }

        /// <summary>
        /// Disposes of the managed resources.
        /// </summary>
        public void Dispose()
        {
            sessionJsonImpl?.Dispose();
        }
    }
}
